barretenberg/cpp/src/barretenberg/protogalaxy/protogalaxy_verifier.cpp

```diff
@@ -35,123 +36,106 @@ void ProtogalaxyVerifier_<DeciderVerificationKeys>::run_oink_verifier_on_each_in
-        lagranges = { (FF(1) - combiner_challenge) * (FF(2) - combiner_challenge) * inverse_two,
-                      combiner_challenge * (FF(2) - combiner_challenge),
-                      combiner_challenge * (combiner_challenge - FF(1)) * inverse_two };
-    } else if constexpr (DeciderVerificationKeys::NUM == 4) {
-        constexpr FF inverse_six = FF(6).invert();
-        vanishing_polynomial_at_challenge = combiner_challenge * (combiner_challenge - FF(1)) *
-                                            (combiner_challenge - FF(2)) * (combiner_challenge - FF(3));
-        lagranges = { (FF(1) - combiner_challenge) * (FF(2) - combiner_challenge) * (FF(3) - combiner_challenge) *
-                          inverse_six,
-                      combiner_challenge * (FF(2) - combiner_challenge) * (FF(3) - combiner_challenge) * inverse_two,
-                      combiner_challenge * (combiner_challenge - FF(1)) * (FF(3) - combiner_challenge) * inverse_two,
-                      combiner_challenge * (combiner_challenge - FF(1)) * (combiner_challenge - FF(2)) * inverse_six };
-    }
-    static_assert(DeciderVerificationKeys::NUM < 5);
-
-    // TODO(https://github.com/AztecProtocol/barretenberg/issues/881): bad pattern
     auto next_accumulator = std::make_shared<DeciderVK>();
     next_accumulator->verification_key = std::make_shared<VerificationKey>(*accumulator->verification_key);
     next_accumulator->is_accumulator = true;
 
-    size_t commitment_idx = 0;
-    for (auto& expected_vk : next_accumulator->verification_key->get_all()) {
-        size_t vk_idx = 0;
-        expected_vk = Commitment::infinity();
-        for (auto& key : keys_to_fold) {
-            expected_vk = expected_vk + key->verification_key->get_all()[commitment_idx] * lagranges[vk_idx];
-            vk_idx++;
-        }
-        commitment_idx++;
-    }
-
     // Compute next folding parameters
+    const auto [vanishing_polynomial_at_challenge, lagranges] =
+        compute_vanishing_polynomial_and_lagrange_evaluations<FF, NUM_KEYS>(combiner_challenge);
     next_accumulator->target_sum =
         perturbator_evaluation * lagranges[0] + vanishing_polynomial_at_challenge * combiner_quotient_evaluation;
-    next_accumulator->gate_challenges =
+    next_accumulator->gate_challenges = // note: known already in previous round
         update_gate_challenges(perturbator_challenge, accumulator->gate_challenges, deltas);
 
-    // Compute ϕ
-    auto& acc_witness_commitments = next_accumulator->witness_commitments;
-    commitment_idx = 0;
-    for (auto& comm : acc_witness_commitments.get_all()) {
-        comm = Commitment::infinity();
-        size_t vk_idx = 0;
-        for (auto& key : keys_to_fold) {
-            comm = comm + key->witness_commitments.get_all()[commitment_idx] * lagranges[vk_idx];
-            vk_idx++;
-        }
-        commitment_idx++;
+    // // Fold the commitments
```

Comment:
Refactor folding to be more succinct, putting grouping logic in the DeciderVerificationKeys object. Ultimately I think this class should go away (you can shove state it in and it's hard to track state in Pg) and these should just be functions, but for now I keep the pattern and put the functions here.